### PR TITLE
Document which methods on `f64` are precise

### PR DESCRIPTION
**Pull request by tbu-**

### CODE REVIEW

**Comment by tbu-** [comment]
It has surprised me in the past that floating-point calculations executed on macOS can be different than the ones on Linux or Windows. This is why I wanted to clearly document which method's exact results can be relied upon and which cannot.

If this is desired, I'll do the same for `f32`.

**Comment by m-ou-se** [comment]
@workingjubilee This seems like something you might have an opinion on?

File: library/std/src/f64.rs [hunk 1]
```diff
@@ -377,6 +411,10 @@ impl f64 {
     ///
     /// Returns NaN if `self` is a negative number other than `-0.0`.
     ///
+    /// # Platform-specific precision
+    ///
+    /// The precision of this function varies by platform and Rust version.
```
Comment (Amanieu) [inline library/std/src/f64.rs:416]: sqrt is defined by the IEEE standard and should always produce the same result on all platforms.

**Review by Amanieu — COMMENTED** [review]
I don't think "precise" is the correct word to use here because it has a very specific meaning in floating-point arithmetic: it means that the float-encoded result is exactly identical to the mathematical result, which may sometimes not be true due to limited float precision.

A better word might be "consistent" to indicate that the result will be consistent across platforms.

File: library/std/src/f64.rs [hunk 2]
```diff
@@ -251,6 +265,10 @@ impl f64 {
     /// this is not always true, and will be heavily dependant on designing
     /// algorithms with specific target hardware in mind.
     ///
+    /// # Platform-specific precision
+    ///
+    /// The results of this function varies by platform and Rust version.
```
[the remark this answers is not in the document]
Comment (tbu-) [reply library/std/src/f64.rs:270]: What happens on platforms that do not support fma? (Do those even exist?) Does it fail to compile?

File: library/std/src/f64.rs (hunk 1 above)
Comment (tbu-) [reply library/std/src/f64.rs:416]: Interesting, I didn't know that.

File: library/std/src/f64.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Amanieu) [reply library/std/src/f64.rs:270]: It is emulated in software, but should still produce the exact result required by the standard.

**Comment by tbu-** [comment]
> I don't think "precise" is the correct word to use here because it has a very specific meaning in floating-point arithmetic: it means that the float-encoded result is exactly identical to the mathematical result, which may sometimes not be true due to limited float precision.

The word precise is only used on the methods that produce results identical to the mathematical result: `floor`, `ceil`, `round`, `round_ties_even`, `trunc`, `fract` and `abs`.

File: library/std/src/f64.rs [hunk 3]
```diff
@@ -278,6 +296,10 @@ impl f64 {
     /// In other words, the result is `self / rhs` rounded to the integer `n`
     /// such that `self >= n * rhs`.
     ///
+    /// # Platform-specific precision
+    ///
+    /// The precision of this function can vary by platform and Rust version.
```
[the remark this answers is not in the document]
Comment (tbu-) [reply library/std/src/f64.rs]: Fixed.

File: library/std/src/f64.rs [hunk 4]
```diff
@@ -311,6 +333,10 @@ impl f64 {
     /// property `self == self.div_euclid(rhs) * rhs + self.rem_euclid(rhs)`
     /// approximately.
     ///
+    /// # Platform-specific precision
+    ///
+    /// The precision of this function can vary by platform and Rust version.
```
[the remark this answers is not in the document]
Comment (tbu-) [reply library/std/src/f64.rs]: Fixed.

File: library/std/src/f64.rs (hunk 1 above)
Comment (tbu-) [reply library/std/src/f64.rs]: Fixed.

File: library/std/src/f64.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tbu-) [reply library/std/src/f64.rs]: Fixed.

**Pull request merged**